src/utils.ts

```diff
@@ -115,51 +36,58 @@ export function errorHandling(errorDetails: any, label?: string, moreInfo?: stri
             }
             break;
-        }
-        default: {
-            window.showErrorMessage(moreInfo + " " +  errorDetails);
+        default:
+            vscode.window.showErrorMessage(moreInfo + " " +  errorDetails);
             break;
-        }
     }
     return;
 }
 
-/*************************************************************************************************************
- * Refresh Profile and Session
- * @param {sessNode} IZoweTreeNode
- * @param {profile} IProfileLoaded
- *************************************************************************************************************/
-export function refreshTree(sessNode: IZoweTreeNode) {
-    const allProf = Profiles.getInstance().getProfiles();
-    for (const profNode of allProf) {
-        if (sessNode.getProfileName() === profNode.name) {
-            sessNode.getProfile().profile = profNode.profile;
-            const SessionProfile = profNode.profile as ISession;
-            if (sessNode.getSession().ISession !== SessionProfile) {
-                sessNode.getSession().ISession.user = SessionProfile.user;
-                sessNode.getSession().ISession.password = SessionProfile.password;
-                sessNode.getSession().ISession.base64EncodedAuth = SessionProfile.base64EncodedAuth;
-                sessNode.getSession().ISession.hostname = SessionProfile.hostname;
-                sessNode.getSession().ISession.port = SessionProfile.port;
-                sessNode.getSession().ISession.rejectUnauthorized = SessionProfile.rejectUnauthorized;
-            }
-        }
-    }
-    sessNode.collapsibleState = TreeItemCollapsibleState.Collapsed;
+export async function resolveQuickPickHelper(quickpick: vscode.QuickPick<vscode.QuickPickItem>): Promise<vscode.QuickPickItem | undefined> {
+    return new Promise<vscode.QuickPickItem | undefined>(
+        (c) => quickpick.onDidAccept(() => c(quickpick.activeItems[0])));
+}
+
+// tslint:disable-next-line: max-classes-per-file
+export class FilterItem implements vscode.QuickPickItem {
+    constructor(private text: string, private desc?: string) { }
+    get label(): string { return this.text; }
+    get description(): string { if (this.desc) { return this.desc; } else { return ""; } }
+    get alwaysShow(): boolean { return false; }
+}
+
+// tslint:disable-next-line: max-classes-per-file
+export class FilterDescriptor implements vscode.QuickPickItem {
+    constructor(private text: string) { }
+    get label(): string { return this.text; }
+    get description(): string { return ""; }
+    get alwaysShow(): boolean { return true; }
+}
+
+/**
+ * Function to retrieve the home directory. In the situation Imperative has
+ * not initialized it we mock a default value.
+ */
+export function getZoweDir(): string {
+    ImperativeConfig.instance.loadedConfig = {
+        defaultHome: path.join(os.homedir(), ".zowe"),
+        envVariablePrefix: "ZOWE"
+    };
+    return ImperativeConfig.instance.cliHome;
 }
```

Comment:
There may be a few more things that could be moved from extensions.ts to some other file.
And functions that could be broken down even more so that people can unit test specific modules/functions/parts of the code.
Examples:
addZoweSession can beboken down in 3-5 smaller modules.
cleandir -like functions don't necessarily belong in the extension.ts but rather in src/utils.ts 

I will create very specific commits for those items in https://github.com/zowe/vscode-extension-for-zowe/pull/668.
Just in case anyone doesn't agree with them, they can be easily reverted.

Reply:
Thank you!!

And ofc they turned off Sonar after all this...but I am sure the code is better after you have cleaned up extension.ts even more :D